kedro/runner/runner.py

```diff
@@ -242,54 +241,86 @@ def _find_persistent_ancestors(
         ``Node``s.
 
     """
-    ancestor_nodes_to_run = set()
+    initial_nodes_to_run: set[Node] = set()
+
     queue, visited = deque(children), set(children)
     while queue:
         current_node = queue.popleft()
-        if _has_persistent_inputs(current_node, catalog):
-            ancestor_nodes_to_run.add(current_node)
+        impersistent_inputs = _enumerate_impersistent_inputs(current_node, catalog)
+
+        # If all inputs are persistent, we can run this node as is
+        if not impersistent_inputs:
+            initial_nodes_to_run.add(current_node)
             continue
-        for parent in _enumerate_parents(pipeline, current_node):
-            if parent in visited:
+
+        # Otherwise, look for the nodes that produce impersistent inputs
+        for node in _enumerate_nodes_with_outputs(pipeline, impersistent_inputs):
+            if node in visited:
                 continue
-            visited.add(parent)
-            queue.append(parent)
-    return ancestor_nodes_to_run
+            visited.add(node)
+            queue.append(node)
+
+    return initial_nodes_to_run
+
+
+def _enumerate_impersistent_inputs(node: Node, catalog: DataCatalog) -> set[str]:
+    """Enumerate impersistent input Datasets of a ``Node``.
+
+    Args:
+        node: the ``Node`` to check the inputs of.
+        catalog: the ``DataCatalog`` of the run.
 
+    Returns:
+        Set of names of impersistent inputs of given ``Node``.
+
+    """
+    # We use _data_sets because they pertain parameter name format
+    catalog_datasets = catalog._datasets
```

Comment:
> @noklam Do you mean here you don't like the terminology of the word impersistent or do you think the implementation itself is leaky and should be changed?

> If it's about terminology, I do agree that impersistent is a tricky word. It's of course the correct opposite of persistent, but for a non-native speaker (me) it makes it harder to read the code for some reason. Maybe just call it non_persistent? 😅 In any case I think it would be helpful to clarify what we mean with impersistent/non_persistent in the context of Kedro. From reading the code (specifically _enumerate_impersistent_inputs), I interpret the definition being anything that is a MemoryDataset. Even if it's defined in the catalog.

@merelcht Mostly terminology, it's just hard to remember as previously it's just MemoryDataset, now is basically memoryDataset + params.

New thought: https://github.com/kedro-org/kedro/pull/3520 - any chance we can utilise the new flag that we added? Can we load `parameters` as a "special" kind of MemoryDataset that will have `self._EPHEMERAL = False`? We can take this PR first and refactor this afterward if needed.

Reply:
`catalog.datasets` members need to be accessible as attributes, so colons and periods are replaced with `__`.  I agree this is very far from an elegant solution, so I'm definitely open to suggestions